### PR TITLE
feat: add sunshine setup

### PR DESCRIPTION
**Pull request by maisatanel**
Sunshine is a host server for streaming applications, kinda like Steam Link but with less latency. It is used via the client Moonlight.
Installation docs: [https://docs.lizardbyte.dev/projects/sunshine/en/latest/about/installation.html#linux](https://docs.lizardbyte.dev/projects/sunshine/en/latest/about/installation.html#linux)
To make it work with Wayland, Sunshine has to be run as root. To make it work in Steam Game Mode, it has to be autostarted.
What I've done here:
- Regrettably, `pulseaudio-utils` has to be layered, because Sunshine uses `pactl`.
- I adapted the postinstall script ([upstream](https://github.com/LizardByte/Sunshine/blob/master/packaging/linux/flatpak/scripts/additional-install.sh)) to create a system service so that the app runs as root. Udev rules aren't installed since they're part of ublue anyway.

Issues:
- The systemd service won't autostart on boot. I don't know why, it starts just fine. Autostart is crucial for getting it to work with Steam Game Mode.

### CODE REVIEW

**Comment by Linwer** [comment]
Alternatively you can just layer the rpm which works flawlessly.

**Comment by Nitrousoxide** [comment]
For the autostart on boot, I've not had a ton of luck with getting it to start after the DE is fully loaded using systemd.  However, someone on their discord suggested this workaround and it seems to be doing okay for me.

Navigate to the auto-start .desktop file location at: 
`/etc/xdg/autostart/`
add a `sunshine.desktop` file with this in it:

```
[Desktop Entry]
Name=Sunshine Server
Type=Application
Exec=bash -c 'sleep 5 && sunshine'
Comment=Sunshine server autostart
```

Now it should start 5 seconds after the user logs in, and the other graphical apps are able to be loaded.  This should prevent it from starting prematurely, though it won't have the crash recovery you get with a proper systemd managed start.

This might even work with the flatpak since that was completely failing to start for folks.  But now that I got the rpm working I don't really feel like messing around with that again.

**Comment by Nitrousoxide** [comment]
Upping the sleep from 5 seconds to 30 in the .desktop file seems to have fixed the terrible streaming performance

```
[Desktop Entry]
Name=Sunshine Server
Type=Application
Exec=bash -c 'sleep 30 && sunshine'
Comment=Sunshine server autostart
```

This is the final version I'm using.

I'm not seeing any hardware support for amd gpu encoding though in base Fedora.  Currently it seems to be using software encoding for systems with AMD gpu's.  Don't know if there's anything else you need to add to the distro to get that working.

**Comment by KyleGospo** [comment]
> I'm not seeing any hardware support for amd gpu encoding though in base Fedora

Sunshine doesn't seem to support VAAPI for encoding, but it does support AMF which is part of AMDGPU-Pro

This is the long way around, but if we want AMF working perhaps the best solution is to create a new Arch image with AMDGPU-Pro and Sunshine installed, then add that to distrobox and auto-launch in gamescope-session as @maisatanel mentioned. That'll get us the streaming performance we want and keep from layering a ton of proprietary stuff on the host.

**Comment by Vitorvlv** [comment]
> Sunshine doesn't seem to support VAAPI for encoding

It does. Not sure if you guys are talking specifically about the flatpak package, but the rpm at least on my fedora 38 desktop (AMD gpu) is doing VAAPI encoding.

AMF support in Sunshine is Windows only. https://github.com/LizardByte/Sunshine/discussions/1331

**Comment by trashbus99** [comment]
> Upping the sleep from 5 seconds to 30 in the .desktop file seems to have fixed the terrible streaming performance
> 
> ```
> [Desktop Entry]
> Name=Sunshine Server
> Type=Application
> Exec=bash -c 'sleep 30 && sunshine'
> Comment=Sunshine server autostart
> ```
> 
> This is the final version I'm using.
> 
> I'm not seeing any hardware support for amd gpu encoding though in base Fedora. Currently it seems to be using software encoding for systems with AMD gpu's. Don't know if there's anything else you need to add to the distro to get that working.

image

**Comment by twaananen** [comment]
Would this work with steam installed in bazzite-arch distrobox?
I tried installing the sunshine flatpack myself and got it to stream the desktop, but I couldn't find out how to get it to launch bazzite-arch steam.

**Comment by EyeCantCU** [comment]
I believe this implementation should be fine as it is what is in the current documentation. One thing I have noticed though is it does not seem to like Intel va-api on first glance. Can do better testing later. Looks like current builds are failing, but as soon as things align with upstream it'll land in unstable. If we need a workaround for wayland (assuming their service configuration doesn't work with it), that shouldn't be too hard to implement and I can explore that in a bit

**Comment by Nitrousoxide** [comment]
So, good news.  I was able to get sunshine working in a (rootful) Ubuntu distrobox container on wayland using the .deb from the project's github.

Here are the steps I took:
```
distrobox-create --nvidia --root --image quay.io/toolbx-images/ubuntu-toolbox:22.04 -n ubuntu -Y
distrobox enter --root ubuntu
(in the ubuntu container)
wget https://github.com/LizardByte/Sunshine/releases/download/v0.21.0/sunshine-debian-bookworm-amd64.deb
sudo apt install -f ./sunshine-debian-bookworm-amd64.deb
```

at this point it should start up and find encoders fine.  I tried streaming to my phone while on Wayland and it worked.

Some concerns:
Being in a container, it won't have access to the hosts Steam or [gnome-randr-rust](https://github.com/maxwellainatchi/gnome-randr-rust) to change resolutions or play the games.  These may need to be symlinked into the container via distrobox-host-exec command as described [here](https://distrobox.it/usage/distrobox-host-exec/).

I was built gnome-randr-rust inside of the container and it worked fine.  I have not had a chance to mess around with trying to get steam on the host to play nice with the container's sunshine.

One final issue is that with this being a rootful distrobox container starting it and entering it may require the user to put in their sudo password.  I'm not sure if that can be avoided using the systemd --system service rather than a systemd --user service.  I know distrobox recommends avoiding using sudo when creating a container, but don't know if this is a concern here.

**Comment by EyeCantCU** [comment]
> So, good news.  I was able to get sunshine working in a (rootful) Ubuntu distrobox container on wayland using the .deb from the project's github.
> 
> Here are the steps I took:
> ```
> distrobox-create --nvidia --root --image quay.io/toolbx-images/ubuntu-toolbox:22.04 -n ubuntu -Y
> distrobox enter --root ubuntu
> (in the ubuntu container)
> wget https://github.com/LizardByte/Sunshine/releases/download/v0.21.0/sunshine-debian-bookworm-amd64.deb
> sudo apt install -f ./sunshine-debian-bookworm-amd64.deb
> ```
> 
> at this point it should start up and find encoders fine.  I tried streaming to my phone while on Wayland and it worked.
> 
> Some concerns:
> Being in a container, it won't have access to the hosts Steam or [gnome-randr-rust](https://github.com/maxwellainatchi/gnome-randr-rust) to change resolutions or play the games.  These may need to be symlinked into the container via distrobox-host-exec command as described [here](https://distrobox.it/usage/distrobox-host-exec/).
> 
> I was able to build gnome-randr-rust inside of the container and it worked fine.  I have not had a chance to mess around with trying to get steam on the host to play nice with the container's sunshine.
> 
> One final issue is that with this being a rootful distrobox container starting it and entering it may require the user to put in their sudo password.  I'm not sure if that can be avoided using the systemd --system service rather than a systemd --user service.  I know distrobox recommends avoiding using sudo when creating a container, but don't know if this is a concern here.
> 
> Edit:
> Doing:
> `sudo ln -s /usr/bin/distrobox-host-exec /usr/bin/steam` did link steam into the container and let sunshine start it up.  though it doesn't show up in the container's process list so "kill steam" doesn't seem to be able to close it.
> Using `steam -shutdown` instead seems to work okay though so that's probably not a huge concern.

Glad you've got something working! Unfortunately, I think it's best we avoid the usage of rootful containers and package the RPM ourselves at this point

**Comment by TNTRedstone** [comment]
I can confirm it works for me on Fedora 39 on Hyprland with NVIDIA proprietary

**Comment by Cubicle13** [comment]
I'm trying to run sunshine on a UM790 Pro (AMD Ryzen 9 iGPU) with ChimeraOS (GNOME using Wayland). Will this work for my system as wel?

**Comment by Nitrousoxide** [comment]
> How do I do this for my steam deck?

This is just for the streaming server.  If you are wanting to stream from a desktop **to** your steamdeck you need to install moonlight on your deck.



> I'm trying to run sunshine on a UM790 Pro (AMD Ryzen 9 iGPU) with ChimeraOS (GNOME using Wayland). Will this work for my system as wel?

It should.  I actually have an amd gpu as well.  I just set the --nvidia flag there to give the distrobox container there more or less universal applicability.  It works fine on my Radeon RX 6600.

**Comment by Cubicle13** [comment]
@Nitrousoxide Could you explain the steps in more detail? I'm trying to apply what you did to the system mentioned before, but I'm very new to this

**Comment by Nitrousoxide** [comment]
> @Nitrousoxide Could you explain the steps in more detail? I'm trying to apply what you did to the system mentioned before, but I'm very new to this

This one-liner *should* do everything I did above (except install gnome-randr-rust for controlling the resolution in wayland)

```
distrobox-create --nvidia --root --pull --image quay.io/toolbx-images/ubuntu-toolbox:22.04 -n sunshine-ubuntu -Y --init-hooks "wget -P /tmp https://github.com/LizardByte/Sunshine/releases/download/v0.21.0/sunshine-debian-bookworm-amd64.deb && apt install -y -f /tmp/sunshine-debian-bookworm-amd64.deb && ln -s /usr/bin/distrobox-host-exec /usr/bin/steam
```
then just `distrobox enter --root  sunshine-ubuntu` and `distrobox-export --app Sunshine` and bam you're done.
If you need the binary too and not just the startmenu app you can also `distrobox-export --bin /usr/bin/sunshine --export-path ~/.local/bin`

You can probably integreate the distrobox-export into that distrobox create command one-liner, but it's gonna require some sudo use in there to change off root since distrobox doesn't like doing app exports as root in the container.

**Comment by Cubicle13** [comment]
@Nitrousoxide meet.google.com/iqk-xyxo-orr

**Comment by Nitrousoxide** [comment]
> @Nitrousoxide Could you explain the steps in more detail? I'm trying to apply what you did to the system mentioned before, but I'm very new to this

I didn't really feel like explaining it all, so I just had chatgpt write it out for me.  This more or less accurately describes what's going on.

```
This command is a complex set of instructions for creating a containerized environment using distrobox-create, a tool that allows you to create and manage Linux containers. Let's break down the command and its options:

    distrobox-create: This is the main command used to create a new container with Distrobox, which is a tool that lets you run different Linux distributions in an isolated environment on your current Linux system.

    --nvidia: This option enables support for Nvidia GPUs within the container. It's useful if you need to run applications that require Nvidia's proprietary drivers or CUDA toolkit.

    --root: This allows the container to run with root privileges. Running a container as root can give the container more access to the host system, which can be necessary for certain applications but also increases security risks.

    --pull: This forces the tool to pull the specified image from the container registry, even if it's already available locally.

    --image quay.io/toolbx-images/ubuntu-toolbox:22.04: This specifies the container image to be used. In this case, it's pulling an Ubuntu Toolbox image (version 22.04) from the quay.io container registry.

    -n sunshine-ubuntu: Sets the name of the container to sunshine-ubuntu.

    -Y: This might be a shortcut for a yes-to-all option, automatically confirming any prompts that might appear during the container creation process.

    --init-hooks: This option allows the execution of custom commands or scripts during the initialization of the container. In this case, it's running a series of commands:
        wget -P /tmp https://github.com/LizardByte/Sunshine/releases/download/v0.21.0/sunshine-debian-bookworm-amd64.deb: Downloads a .deb package (Sunshine) from GitHub and saves it to the /tmp directory inside the container.
        apt install -y -f /tmp/sunshine-debian-bookworm-amd64.deb: Installs the downloaded package using apt. The -y flag automatically answers 'yes' to prompts, and -f attempts to fix broken dependencies if needed.
        ln -s /usr/bin/distrobox-host-exec /usr/bin/steam: Creates a symbolic link from distrobox-host-exec to steam. This could be a way to trick software into thinking it's running in a normal environment when it's actually inside a container.

In summary, this command creates a new Ubuntu 22.04 container with Nvidia support and root privileges, installs the Sunshine application (presumably for game streaming or similar purposes), and sets up an environment that mimics a typical Linux desktop, possibly for better integration with certain applications like Steam.
```

**Comment by Cubicle13** [comment]
@Nitrousoxide It WORKS!
Uninstalled chimeraOS and installed base Bazzite
Ctl alt T and entered the commands from your earlier post one by one.

**Comment by Nitrousoxide** [comment]
> > So, good news.  I was able to get sunshine working in a (rootful) Ubuntu distrobox container on wayland using the .deb from the project's github.
> > Here are the steps I took:
> > ```
> > distrobox-create --nvidia --root --image quay.io/toolbx-images/ubuntu-toolbox:22.04 -n ubuntu -Y
> > distrobox enter --root ubuntu
> > (in the ubuntu container)
> > wget https://github.com/LizardByte/Sunshine/releases/download/v0.21.0/sunshine-debian-bookworm-amd64.deb
> > sudo apt install -f ./sunshine-debian-bookworm-amd64.deb
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > at this point it should start up and find encoders fine.  I tried streaming to my phone while on Wayland and it worked.
> > Some concerns:
> > Being in a container, it won't have access to the hosts Steam or [gnome-randr-rust](https://github.com/maxwellainatchi/gnome-randr-rust) to change resolutions or play the games.  These may need to be symlinked into the container via distrobox-host-exec command as described [here](https://distrobox.it/usage/distrobox-host-exec/).
> > I was able to build gnome-randr-rust inside of the container and it worked fine.  I have not had a chance to mess around with trying to get steam on the host to play nice with the container's sunshine.
> > One final issue is that with this being a rootful distrobox container starting it and entering it may require the user to put in their sudo password.  I'm not sure if that can be avoided using the systemd --system service rather than a systemd --user service.  I know distrobox recommends avoiding using sudo when creating a container, but don't know if this is a concern here.
> > Edit:
> > Doing:
> > `sudo ln -s /usr/bin/distrobox-host-exec /usr/bin/steam` did link steam into the container and let sunshine start it up.  though it doesn't show up in the container's process list so "kill steam" doesn't seem to be able to close it.
> > Using `steam -shutdown` instead seems to work okay though so that's probably not a huge concern.
> 
> Glad you've got something working! Unfortunately, I think it's best we avoid the usage of rootful containers and package the RPM ourselves at this point

I noticed that they do have a fedora 39 rpm on their nightly, and it seems to work for me.
https://github.com/LizardByte/Sunshine/releases/tag/nightly-dev

**Comment by KyleGospo** [comment]
https://github.com/ublue-os/bazzite/commit/ed975d1601e4f0c0305efe5da8ad3049549d317d

This should make things easier for everyone.

**Comment by hellomp** [comment]
> > @Nitrousoxide Could you explain the steps in more detail? I'm trying to apply what you did to the system mentioned before, but I'm very new to this
> 
> This one-liner _should_ do everything I did above (except install gnome-randr-rust for controlling the resolution in wayland)
> 
> ```
> distrobox-create --nvidia --root --pull --image quay.io/toolbx-images/ubuntu-toolbox:22.04 -n sunshine-ubuntu -Y --init-hooks "wget -P /tmp https://github.com/LizardByte/Sunshine/releases/download/v0.21.0/sunshine-debian-bookworm-amd64.deb && apt install -y -f /tmp/sunshine-debian-bookworm-amd64.deb && ln -s /usr/bin/distrobox-host-exec /usr/bin/steam
> ```
> 
> then just `distrobox enter --root sunshine-ubuntu` and `distrobox-export --app Sunshine` and bam you're done. If you need the binary too and not just the startmenu app you can also `distrobox-export --bin /usr/bin/sunshine --export-path ~/.local/bin`
> 
> You can probably integreate the distrobox-export into that distrobox create command one-liner, but it's gonna require some sudo use in there to change off root since distrobox doesn't like doing app exports as root in the container.

Just to see if I got it right. This is not supposed to work on Game Mode, right? It worked great on Desktop Mode for me, but it seems that the process is closed when I switch to Game Mode

**Comment by Nitrousoxide** [comment]
> > > @Nitrousoxide Could you explain the steps in more detail? I'm trying to apply what you did to the system mentioned before, but I'm very new to this
> > 
> > 
> > This one-liner _should_ do everything I did above (except install gnome-randr-rust for controlling the resolution in wayland)
> > ```
> > distrobox-create --nvidia --root --pull --image quay.io/toolbx-images/ubuntu-toolbox:22.04 -n sunshine-ubuntu -Y --init-hooks "wget -P /tmp https://github.com/LizardByte/Sunshine/releases/download/v0.21.0/sunshine-debian-bookworm-amd64.deb && apt install -y -f /tmp/sunshine-debian-bookworm-amd64.deb && ln -s /usr/bin/distrobox-host-exec /usr/bin/steam
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then just `distrobox enter --root sunshine-ubuntu` and `distrobox-export --app Sunshine` and bam you're done. If you need the binary too and not just the startmenu app you can also `distrobox-export --bin /usr/bin/sunshine --export-path ~/.local/bin`
> > You can probably integreate the distrobox-export into that distrobox create command one-liner, but it's gonna require some sudo use in there to change off root since distrobox doesn't like doing app exports as root in the container.
> 
> Just to see if I got it right. This is not supposed to work on Game Mode, right? It worked great on Desktop Mode for me, but it seems that the process is closed when I switch to Game Mode

I don't imagine so.  Many processes are killed for gamemode.
